File: Crio/Recursion/16_KthSmallestElementInABst/Solution.java
import crio.ds.Tree.TreeNode;
import java.util.*;

/*
// Definition of TreeNode
public class TreeNode {
    public long val;
    public TreeNode left;
    public TreeNode right;

    public TreeNode (long x) {
        val = x;
        left = null;
        right = null;
    }
}
*/


public class Solution {
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    public static void dfs(TreeNode vertex, List<Long> inOrder) {
        if (vertex == null)
            return;
        dfs(vertex.left, inOrder);
        inOrder.add(vertex.val);
        dfs(vertex.right, inOrder);
    }

    public static List<Long> inorderTraversal(TreeNode root) {
        List<Long> inOrder = new ArrayList<Long>();
        dfs(root, inOrder);
        return inOrder;
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    public int kthSmallestElementInABst(TreeNode root, int k) {
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        List<Long> inOrder = inorderTraversal(root);
        return Math.toIntExact(inOrder.get(k - 1));
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
}

File: Crio/Recursion/16_KthSmallestElementInABst/Solution.cpp
#include "../crio/cpp/ds/TreeNode/TreeNode.hpp"

// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
void fun(TreeNode* root, vector<int>& arr) {
    if (root == NULL)
        return;
    fun(root->left, arr);
    arr.push_back(root->val);
    fun(root->right, arr);
}
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

/*
// Definition for TreeNode
class TreeNode {
public:
    long val;
    TreeNode* left;
    TreeNode* right;
    TreeNode (long x) {
        val = x;
        left = NULL;
        right = NULL;
    }
};
*/
int kthSmallestElementInABst(TreeNode* root, int k) {
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    vector<int> arr;
    fun(root, arr);
    return arr[k - 1];
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}
